benchmark/single-source/BucketSort.swift

```diff
@@ -0,0 +1,139 @@
+// array of generic type of "SortableItem" to be sorted. Unfortunately,  if
+// ``sortingAlgo'' type is not known for the callsite in line 84, then the 
+// ``sort'' method can not be specialized to integer array sorting, which will 
+// lead to a huge performance loss. Since SortingAlgorithm and InsertionSort are 
+// declared to be ``public'' and that lines 83-85 can not be inlined in
+// BucketSortImpl (due to inlining heuristic limitations), today swift 
+// compiler (without ExistentialSpecializer) can not achieve this feat.  With 
+// ExistentialSpecializer which enables generic specialization recursively in a
+// call chain, we are able to specialize line 84 for InsertionSort on integers.
+
+import TestsUtils
+import Foundation
+
+public let BucketSort = BenchmarkInfo(
+  name: "BucketSort",
+  runFunction: run_BucketSort,
+  tags: [.validation, .algorithm],
+  setUpFunction: { buildWorkload() },
+  legacyFactor: 10)
+
+public protocol IntegerConvertible {
+    func convertToInt() -> Int
+}
+extension Int: IntegerConvertible, SortableItem {
+    public func convertToInt() -> Int {
+        return self
+    }
+}
+public protocol SortableItem: IntegerConvertible, Comparable { }
+public protocol SortingAlgorithm {
+    func sort<T: SortableItem>(_ items: [T]) -> [T]
+}
+public struct InsertionSort: SortingAlgorithm {
+    public func sort<T: SortableItem>(_ items: [T]) -> [T] {
+        var sortedItems = items
+        for i in 0 ..< sortedItems.count {
+            var j = i
+            while j > 0 && sortedItems[j-1] > sortedItems[j] {
+                let temp = sortedItems[j-1]
+                sortedItems[j-1] = sortedItems[j]
+                sortedItems[j] = temp
+                j -= 1
+            }
+        }
+        return sortedItems
+    }
+}
+func distribute<T>(_ item: T, bucketArray: inout [Bucket<T>]) {
+  let val = item.convertToInt()
+  let capacity = bucketArray.first!.capacity
+  let index = val / capacity
+  bucketArray[index].add(item)
+}
+struct Bucket<T: SortableItem> {
+  var items: [T]
+  let capacity: Int
+  init(capacity: Int) {
+    self.capacity = capacity
+    items = [T]()
+  }
+  mutating func add(_ item: T) {
+    if items.count < capacity {
+      items.append(item)
+    }
+  }
+  func sort(_ sortingAlgo: SortingAlgorithm) -> [T] {
+    return sortingAlgo.sort(items)
+  }
+}
+func BucketSortImpl<T>(_ items: [T], sortingAlgorithm: SortingAlgorithm, bucketArray: [Bucket<T>]) -> [T] {
+  var copyBucketArray = bucketArray
+  for item in items {
+    distribute(item, bucketArray: &copyBucketArray)
+  }
+  var sortedArray = [T]()
+  for bucket in copyBucketArray {
+    sortedArray += bucket.sort(sortingAlgorithm)
+  }
+  return sortedArray
+}
+func isArraySorted(_ arr: [Int]) -> Bool {
+  var idx = 0
+  while idx < (arr.count - 1) {
+    if arr[idx] > arr[idx+1] {
+       return false
+    }
+    idx += 1
+  }
+  return true
+}
+let NUMITEMS = 2500
```

Comment:
@atrick reduced this from 10000 to 2500. should be fine now?

Reply:
Sorry, my fault! The more important run time will be after you enable ExistentialSpecialization. I did this locally, and it only takes 160µs with 10,000 items. So that was actually a better choice.

The only warning I see now is this.
```
memory: 'BucketSort' has very wide range of memory used between independent, repeated measurements.
memory: 'BucketSort' mem_pages [i1, i2]: min=[19, 19] 𝚫=0 R=[0, 24]
```

If you think of a way to improve it great, but I don't think it needs to hold up the commit.

Reply:
That memory warning is probably too sensitive. This benchmark looks just fine as is from the memory usage perspective.